test/test_quantized.py

```diff
@@ -121,5 +171,81 @@ def test_qsumrelu_different_qparams(self):
         np.testing.assert_equal(qC, qC_hat.int_repr())
 
 
-if __name__ == '__main__':
+@unittest.skipIf(
+    TEST_WITH_UBSAN or not torch.fbgemm_is_cpu_supported(),
+    " Quantized FC requires FBGEMM. FBGEMM does not play"
+    " well with UBSAN at the moment, so we skip the test if"
+    " we are in a UBSAN environment.",
+)
+class TestQuantizedFC(unittest.TestCase):
+    """Tests the correctness of the quantized::fc op."""
+
+    def test_qfc(self):
+        qfc_prepack = torch.ops.quantized.fbgemm_linear_prepack
+        qfc = torch.ops.quantized.fbgemm_linear
+
+        batch_size = 4
+        input_channels = 16
+        output_channels = 8
+
+        X_scale = 1.5
+        X_zp = 5
+        X_value_min = 0
+        X_value_max = 225
+        X_q = np.round(
+            np.random.rand(batch_size, input_channels) * (X_value_max - X_value_min)
+            + X_value_min
+        ).astype(np.uint8)
+
+        W_scale = 0.4
+        W_zp = 2
+        W_value_min = -128
+        W_value_max = 127
+        W_q = np.round(
+            np.random.rand(output_channels, input_channels)
+            * (W_value_max - W_value_min)
+            + W_value_min
+        ).astype(np.int8)
+
+        b_q = np.round(np.random.randn(output_channels) * 10 - 10).astype(np.int32)
+
+        # Make sure X_scale * W_scale * input_channels * X_value_max * W_value_max
+        # matches Y_scale * 255 (max for uint8).
+        Y_scale = 1345.1234
+        Y_zp = 5
+
+        avoid_vpmaddubsw_overflow_fc(
+            batch_size,
+            input_channels,
+            output_channels,
+            X_q,
+            X_value_min,
+            X_value_max,
+            W_q,
+            W_value_min,
+            W_value_max,
+        )
+
+        # Reference quantized FC operator
+        Y_q_ref = qfc_ref(X_q, X_scale, X_zp, W_q, W_scale, W_zp, b_q, Y_scale, Y_zp)
```

Comment:
Will be great to have a test that checks a floating point fc reference operator with quantized weights, activations and biases and ensure that the output matches. Test would be something like:
W_f = W_q.dequantize()
X_f = X_q.dequantize()
Y_f = F.Linear(W_f,X_f,B_f)
assert(Y_f==Y_q_ref.dequantize())

Reply:
Due to the clamp function in the requantization from int32_t to uint_t (out[i] = clamp(in[i] * scale + zero_point, 0, 255)), there should be accuracy loss. Then Y_f should not be equal to Y_q_ref.dequantize(). Correct me if I am wrong.